File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.py
import sys
input = sys.stdin.readline
import itertools
import copy

def rotate(fr_row, to_row, fr_col, to_col):
  this_turn_arr = []
  for i in range(fr_col, to_col):
    this_turn_arr.append(this_turn_matrix[fr_row][i])
  for i in range(fr_row, to_row):
    this_turn_arr.append(this_turn_matrix[i][to_col])
  for i in reversed(range(fr_col+1, to_col+1)):
    this_turn_arr.append(this_turn_matrix[to_row][i])
  for i in reversed(range(fr_row+1, to_row+1)):
    this_turn_arr.append(this_turn_matrix[i][fr_col])
  idx = -1
  for i in range(fr_col, to_col):
    this_turn_matrix[fr_row][i] = this_turn_arr[idx]
    idx+=1
  for i in range(fr_row, to_row):
    this_turn_matrix[i][to_col] = this_turn_arr[idx]
    idx+=1
  for i in reversed(range(fr_col+1, to_col+1)):
    this_turn_matrix[to_row][i] = this_turn_arr[idx]
    idx += 1
  for i in reversed(range(fr_row+1, to_row+1)):
    this_turn_matrix[i][fr_col] = this_turn_arr[idx]
    idx += 1
  

n, m, k = map(int,input().split())

matrix = [list(map(int,input().split())) for _ in range(n)]

proc = [list(map(int,input().split())) for _ in range(k)]

result = float('inf')

for arr in itertools.permutations(range(k)):
  this_turn_matrix = copy.deepcopy(matrix)
  for idx in arr:
    this_row, this_col, rotate_len = proc[idx][0]-1, proc[idx][1]-1, proc[idx][2]
    for j in range(1, rotate_len+1):
      rotate(this_row-j, this_row + j, this_col - j, this_col + j)
  min_row = float('inf')
  for i in range(n):
    min_row = min(min_row, sum(this_turn_matrix[i]))
  result = min(result,min_row)
print(result)

File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int n,m, k;
    static int[][] matrix;
    static int[][] proc;
    static int[] result;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int maxSum = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        matrix = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                matrix[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        proc = new int[k][3];
        for (int t = 0; t < k; t++) {
            st = new StringTokenizer(br.readLine());
            int row = Integer.parseInt(st.nextToken()) - 1;
            int col = Integer.parseInt(st.nextToken()) - 1;
            int length = Integer.parseInt(st.nextToken());
            proc[t][0] = row;
            proc[t][1] = col;
            proc[t][2] = length;

        }
        result = new int[k];
        dfs(0, 0);
        System.out.println(maxSum);
    }

    private static void dfs(int index, int bit) {
        if (index == k) {
            int[][] copyMatrix = new int[n][m];
            for (int i = 0; i < n; i++) {
                copyMatrix[i] = matrix[i].clone();

            }

            for (int idx : result) {
                rotate(proc[idx]);
            }

            for (int i = 0; i < n; i++) {
                maxSum = Math.min(maxSum,Arrays.stream(matrix[i]).sum());
            }
            matrix = copyMatrix;
            return;
        }
        for (int i = 0; i < k; i++) {
            if ((bit & (1 << i)) == 0) {
                result[index] = i;
                dfs(index + 1, bit | (1 << i));
            }
        }
    }

    private static void rotate(int[] ints) {
        int thisRow = ints[0];
        int thisCol = ints[1];
        int length = ints[2];
        for (int i = 1; i <= length; i++) {

            int direction = 0;
            Queue<Integer> queue = new ArrayDeque<>();

            int startRow = thisRow - i;
            int startCol = thisCol - i;
            for (int p = 0; p < 4; p++) {
                for (int k = 0; k < 2 * i; k++) {
                    queue.add(matrix[startRow][startCol]);
                    startRow += dx[direction];
                    startCol += dy[direction];
                }
                direction += 1;
            }



            startRow = thisRow - i;
            startCol = thisCol - i + 1;
            direction = 0;

            for (int k = 1; k < 2 * i; k++) {
                matrix[startRow][startCol] = queue.poll();
                startRow += dx[direction];
                startCol += dy[direction];
            }
            direction += 1;
            for (int p = 0; p < 3; p++) {
                for (int k = 0; k < 2 * i; k++) {
                    matrix[startRow][startCol] = queue.poll();
                    startRow += dx[direction];
                    startCol += dy[direction];
                }
                direction += 1;
            }
//            System.out.println(startCol);
            matrix[startRow][startCol] = queue.poll();


//            for (int r = 0; r < n; r++) {
//                for (int c = 0; c < m; c++) {
//                    System.out.print(matrix[r][c] + " ");
//
//                }
//                System.out.println();
//            }
//            System.out.println();
        }

    }
}
